### PR TITLE
fix(profiles): isolate Hermes home per streaming turn

### PR DESCRIPTION
**Pull request by starship-s**
## Thinking Path

- Concurrent chats can resolve different Hermes profiles inside the same WebUI process.
- The streaming path resolved the correct profile home, but agent construction and later prompt/config/skill reads still depended on process-global `HERMES_HOME` and globally patched skill-module paths after the environment lock was released.
- A second profile turn could overwrite those globals before the first turn finished constructing its agent, allowing profile-specific prompt, config, or skill state to cross the session boundary.
- Bind the resolved home through Hermes Agent's context-local override for the complete streaming turn and detached profile-worker scope, then reset it unconditionally during teardown.
- Keep environment mirroring for legacy/subprocess readers; mutate skill-module globals only when the override is unavailable or either loaded module lacks call-time profile-home resolution, and serialize that static-module fallback for the full turn.

## What Changed

- Install the profile-home context override before agent construction, MCP discovery, prompt loading, and tool execution, and reset it on every streaming exit path.
- Capability-check both loaded skill modules under the active override; skip global path patching only when both resolve the request profile dynamically.
- Apply the same isolation rule to profile-scoped background workers, preserving snapshot/patch/restore for static modules and serializing only that compatibility path.
- Keep checkpoint-only `Session.save()` profile-scoped without entering skill-module fallback locking, avoiding checkpoint/final-writeback lock inversion.
- Add deterministic concurrent-profile coverage using real skill discovery, explicit teardown coverage, and legacy-fallback regression coverage.
- Update existing title, compression, and update-summary tests to assert context-local isolation rather than global cache mutation.

## Why It Matters

- Prevents one profile's `SOUL.md`, skills, or config from influencing another profile's chat during concurrent execution.
- Keeps a contaminated agent from being cached under an otherwise-correct session identity.
- Preserves compatibility with older Hermes Agent versions while making current-agent profile isolation request-local and exception-safe.
- This is security-sensitive environment handling: the regression suite deliberately clobbers process-global profile state mid-run and verifies each request still resolves only its own home.

## Contract Routing

Task type: runtime/profile isolation fix.

Touched areas:
- streaming turn setup and teardown
- detached profile-worker environment scope
- profile-routing regression tests

Relevant public docs:
- `AGENTS.md`
- `CONTRIBUTING.md`
- `docs/CONTRACTS.md`
- `docs/rfcs/webui-run-state-consistency-contract.md` (invariant 8: name the mutated layer and prove the invariant)

Scope boundaries:
- no SSE event-shape, transcript, replay, sidebar, API, or UI changes
- no new dependency or persistent state format

Evidence:
- deterministic overlapping profile scopes with opposite process-global homes and real `skills_list()` reads
- explicit override teardown and legacy-fallback coverage

## Verification

Focused profile/runtime coverage:

`PYTHONPATH=/path/to/hermes-agent python -m pytest -o addopts= -q tests/test_issue2023_hermes_home_skill_modules.py tests/test_issue2848_checkpoint_profile.py tests/test_issue5567_profile_home_override.py tests/test_title_aux_routing.py tests/test_sprint46.py tests/test_update_banner_fixes.py`

Result: `198 passed, 8 subtests passed`

Adjacent environment/profile coverage:

`python -m pytest -o addopts= -q --timeout=60 tests/test_profile_path_security.py tests/test_sprint29.py::TestENVLock tests/test_issue2024_env_lock_skill_imports.py tests/test_sprint42.py tests/test_issue1164_env_file_corruption.py tests/test_issue1362_codex_oauth_onboarding.py`

Result: `80 passed`

Race regressions: checkpoint lock contention, patched-global alpha/beta overlap with real skill discovery, static-module fallback serialization, and dynamic context-local skill isolation each passed `10/10` repeated runs on the final candidate.

Full repository suite:

`./scripts/test.sh`

Result: `13,370 passed`; pristine current `master` produced `13,356 passed` and the exact same six local environment/order-dependent failures (`0` branch-introduced failures).

Static checks:

- `python scripts/ruff_lint.py --diff origin/master` — 0 findings on changed lines
- `python -m py_compile ...` — passed for all eight changed files
- `git diff --check origin/master...HEAD` — passed

## Risks / Follow-ups

- Hermes Agent versions without the context-local API, plus v0.18.0-style static skill modules, use a full-turn serialized snapshot/patch/restore fallback; dynamic-capable versions remain concurrent.
- This does not redesign other process-global runtime registries; it narrowly fixes profile-home resolution for foreground streaming and existing profile-scoped worker contexts.
- Release-note wording: concurrent chats now keep profile-specific prompt, config, and skill resolution isolated per turn.

## Model Used

- OpenAI / `gpt-5.3-codex-spark` via OpenCode for implementation.
- OpenAI / `gpt-5.6-sol` for orchestration and verification.
- OpenAI / `gpt-5.3-codex-spark` via Codex for rebase conflict resolution and current-upstream test adaptation.
- Anthropic / `claude-opus-4-8` via Claude Code for independent current-head review (`SIGN-OFF WITH NITS`, no blockers).

### CODE REVIEW

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
Nice work on the per-turn isolation — the concurrency reproduction confirms scoped reads (model, prompt, skills, config) are correctly isolated per streaming turn under overlapping cross-profile turns, and the home override restores on all exits. But the gate found **one version-dependent skill-isolation gap** that's a real cross-profile leak on a specific Agent version:

**[SILENT — cross-profile skill leak on Hermes Agent v0.18.0] `api/streaming.py:7203` + `api/profiles.py:1194`.** The code treats a successful home-override token as proof that the skill modules resolve their paths dynamically at call time, so it **skips the legacy `patch_skill_home_modules()` path-patch**. But **v0.18.0 exposes the override API while both skill modules still hold a static `SKILLS_DIR`** — so on v0.18.0, named-profile chats and background workers silently load the **default profile's skills**.
- **Verified:** with v0.18.0, the base (pre-PR) returned `alpha-only` (correct profile-local skills); this PR returned `default-only` (wrong profile). The full-stream probe passes on v0.18.1/current Agent, so it's version-gated — which is exactly the kind of gap that slips through if you only test against the current Agent.
- **Fix:** don't infer dynamic-resolution capability from the override token alone — **capability-check both skill modules** after prewarming. Skip the patch only when the override install succeeded AND both skill modules genuinely support call-time home resolution; otherwise **retain the snapshot/patch/restore** (`patch_skill_home_modules()`) for v0.18.0/static modules.
- **Test:** cover the "override API present + static skill paths" combination (simulate v0.18.0) and assert profile-local skills, so the version gap can't regress.

The rest is solid — concurrent-turn model/prompt/config/skill scoping isolated, restore on all exit paths, no single-profile regression, 1094 adjacent tests green. Just make the skill-path decision capability-based (not token-based) so v0.18.0 doesn't fall back to default-profile skills. Once that lands I'll re-gate. Thanks @starship-s — this closes a real concurrent-profile isolation hole.

File: api/streaming.py [hunk 1]
```diff
@@ -7119,18 +7192,17 @@ def _agent_status_callback(kind, message):
...
+                # tools.skill_manager_tool._skills_dir() can resolve the active
+                # profile from get_hermes_home() and keep per-thread isolation
+                # without mutating module globals. If the override installation
+                # fails (older agents), retain the legacy process-cache patch
+                # fallback and keep issue-2024 coverage meaningful.
                 if patch_skill_home_modules is not None:
-                    patch_skill_home_modules(Path(_profile_home))
+                    _has_streaming_override = _streaming_hermes_home_override_ctx[1] is not None
+                    if not _has_streaming_override:
+                        patch_skill_home_modules(Path(_profile_home))
```
[the remark this answers is not in the document]
Comment (starship-s) [reply api/streaming.py]: Addressed in `594784a2`. Override installation is now tracked with an explicit boolean independent of the opaque reset token, so a valid `None` token is still reset. Added streaming and background regressions for `None` tokens, plus a capability-probe exception case that verifies a successfully installed override is always cleared.

**Comment by starship-s** [comment]
Addressed the v0.18.0 skill-isolation gate in `594784a2`. The decision is now behavioral rather than token/version based: after prewarming and installing the request-local home, both loaded skill modules must resolve that profile via their callable `_skills_dir()` before global patching is skipped. Static, missing, raising, or wrong-path modules use snapshot/patch/restore guarded by a shared fallback-only lock held for the full worker/stream turn; dynamic-capable Agent versions remain concurrent. Regression coverage includes the override-present + static-module combination, two-profile fallback serialization and exception release, valid `None` reset tokens, capability-probe teardown, and the real dynamic concurrent `skills_list()` path. Final local results: focused `174 passed` (+ 8 subtests), adjacent `79 passed`, three relevant race/teardown nodes `10/10` each, and full suite `12,582 passed` with only the same six pristine-`master` environment failures. Ready for re-gating when CI completes.

**Review by nesquena-hermes — CHANGES_REQUESTED** [review]
Round 2 — the v0.18.0 gap from my last bounce is **verified CLOSED** ✅: a named-profile turn on published v0.18.0 modules now loads `alpha-only` (not `default-only`) with proper restoration, and the v0.18.1 dynamic path resolves `alpha-only` without patching. The capability gate does the right thing on the clean single-turn case for both versions. Version-specific slices 9/9 each, adjacent 79/79.

But the deep concurrency + teardown probe found **2 release-blockers** the capability/lock rework introduced — both in the new machinery:

**1. [CORE — lock-ordering deadlock/hang] `api/streaming.py:8440` (checkpoint save).** `_save_streaming_checkpoint()` takes the skill snapshot/patch lock, then on a long-tool turn times out joining the checkpoint (line 8536) and waits for `_agent_lock` during final writeback — an `_agent_lock` → skill-lock ordering that can hang the turn (no `done`/`stream_end`). **Verified with published v0.18.0 modules.** **Fix:** let `_save_streaming_checkpoint()` use profile-home scoping **without** skill snapshot/patch locking — `Session.save()` doesn't access skills, so the checkpoint save never needs the skill lock. Add a long-tool-turn regression proving `done` + `stream_end` still occur.

**2. [SILENT — concurrency cross-profile skill leak] `api/profiles.py:80` (capability check).** The check can mistake a **globally-patched** `_skills_dir()` for genuine context-local resolution. **Verified with overlapping turns:** with v0.18.1 modules already patched to profile alpha, alpha classifies as "dynamic" and skips the lock, while beta takes the fallback and patches globals → **alpha's real `skills_list()` then returned `beta-only`** (cross-profile skill leak under concurrency). Streaming makes this decision at `api/streaming.py:7192`. **Fix:** only classify both modules as dynamic when `SKILLS_DIR` still equals `_SKILLS_DIR_AT_IMPORT` **and** `_skills_dir()` actually resolves the requested override; otherwise use the serialized snapshot/patch fallback. Add an overlapping alpha/beta regression that starts from a globals-patched-to-alpha state.

The core isolation concept is right and the version gap is genuinely fixed — these two are the concurrency/teardown edges the capability+lock machinery introduced. Once the checkpoint save drops the skill lock (#1) and the capability check rejects globally-patched state as "dynamic" (#2), re-gate. Thanks @starship-s — solid progress on a genuinely tricky isolation problem.

**Comment by starship-s** [comment]
Addressed both current-head blockers in `74cb6d03`. (1) Checkpoint-only `Session.save()` now uses profile scoping with `scope_skill_modules=False`, so it never enters skill capability checks or snapshot/patch locking; a contention regression proves it completes while the fallback lock is held. (2) Capability detection now requires each module global `SKILLS_DIR` to still equal `_SKILLS_DIR_AT_IMPORT` before trusting `_skills_dir()`; a real-module alpha/beta overlap regression starts from globals patched to alpha, proves beta blocks, and verifies profile-correct skill discovery and restoration. Final verification: 178 focused tests + 8 subtests, both new contention/overlap regressions 10/10, full suite 12,586 passed with only the six reproduced upstream/environment failures, Ruff/compile/whitespace clean, and independent Fable `SIGN-OFF`. Please re-gate current head `74cb6d03`.

**Comment by nesquena-hermes** [comment]
## Re-gate (round 3): SAFE TO SHIP ✅ — converged

Verified at head `74cb6d0329` (Codex reproduce + full 12,595-test suite, no regression):

- **Deadlock CLOSED** (`api/streaming.py:5089`): checkpoint saves now use `scope_skill_modules=False`, so a long-tool-turn checkpoint no longer takes the skill snapshot/patch lock — no `_agent_lock`→skill-lock ordering hang. A real v0.18.0 long-tool stream completed checkpointing and emitted `done` before `stream_end`.
- **Concurrency skill-leak CLOSED** (`api/profiles.py:109`): the capability check now rejects globally-patched module state as "dynamic". Repeated overlapping alpha/beta turns (10× race runs) each returned their own `alpha-only` / `beta-only` skills, with correct restoration.
- **Not regressed:** published v0.18.0 static-module + v0.18.1 dynamic paths both resolve profile-local skills; exception / cancellation / teardown paths all restore both context-local and global state; concurrent per-turn home isolation intact. 178 focused + 79 adjacent + full 12,595-suite green.

This closes the concurrent-profile isolation hole (a named-profile turn could read another profile's home/skills under concurrency). Three rounds of deep gating hardened the whole per-turn home-scoping + skill-module path.

⚠️ **HELD for maintainer merge:** profile-isolation (security boundary) + streaming crown-jewel path + L-sized — gate-certified + decision-ready, but not auto-merged (sensitive-path policy). Ready to ship on your go. Thanks @starship-s for the thorough multi-round convergence.

**Comment by nesquena-hermes** [comment]
Status note (no new findings — the round-3 gate-cert above still stands on the code): this branch has since drifted into a **merge conflict with master** (`mergeable: CONFLICTING`), and the conflict is in `api/streaming.py` — the security-boundary file this PR edits — plus `tests/test_sprint46.py` and `tests/test_update_banner_fixes.py`.

So the path to ship is: rebase onto current master (`69ffaf344`) to resolve the conflict, and because the conflict touches the per-turn Hermes-home isolation on `api/streaming.py`, the round-3 cert can't carry across a head change on a security boundary — it needs a **fresh adversarial re-gate on the rebased head** before merge. The isolation design itself reviewed clean at `74cb6d0329` (context-local override + symmetric finally teardown, fail-closed `_SKILL_HOME_MODULE_PATCH_LOCK`); this is purely the mechanical rebase + re-verify, then it's the maintainer's sensitive-path merge call.

Flagging so it doesn't sit silently as CONFLICTING. Thanks @starship-s.

**Comment by starship-s** [comment]
Rebased onto current `master` (`bbb5446b`) and force-pushed current head `8838252b`. Conflict integration preserves upstream `_turn_pending_source`, runtime-refresh, pending/TTFT metering, and teardown behavior alongside the profile-home isolation hardening. The affected streaming test harness now matches current provider/meter interfaces and still asserts exactly one PR-owned fallback snapshot/patch/restore cycle.

Verification on `8838252b`: 198 focused tests + 8 subtests; 80 adjacent profile/env-lock tests; four lock/race regressions passed 10/10 iterations; Ruff reports 0 findings on changed lines; all eight changed Python files compile; diff checks pass. Full suite: 13,370 passed. Pristine `master` under the identical environment: 13,356 passed with the exact same six failures, so branch delta is 0 new failures.

Fresh independent Claude Opus adversarial review of `8838252b` returned `SIGN-OFF WITH NITS` with no blockers. The nits are non-blocking legacy-path throughput and pre-existing structural observations. Please re-gate the current head.

**Comment by nesquena-hermes** [comment]
## 🔬 Gate certification — GREEN ✅
**Certified head:** `sha:8838252bf0e8` · **PR:** #5877 · _@starship-s, fix(profiles): isolate Hermes home per streaming turn_
**Verdict:** Convergence complete. All three prior maintainer blockers are reproduced-closed against the real published v0.18.0/v0.18.1 skill modules, the full suite is green, and the one advisor finding is byte-identical pre-existing master infra (improved, not worsened, by this PR). Engineering certified clean — final merge parks for @nesquena as a profile-isolation / concurrency-sensitive crown-jewel PR.

### What I ran (isolated worktree `/tmp/wt-gate-5877`, patch-id-clean rebase on current `origin/master`)
| Gate | Result |
|------|--------|
| Rebase | Clean 3-way onto current `origin/master` (was 22 behind); rebased net diff **patch-id byte-identical** to the PR head (`f77a11d4…`). Branch `gate-b/5877-rebase` pushed for author-attributed staging. |
| Codex (reproduce) | **SAFE TO SHIP** — reproduced all 3 blockers closed with real v0.18.0/v0.18.1 modules; no `_agent_lock`/skill-lock/`_ENV_LOCK` inversion; no regression. |
| Opus (senior review) | **SAFE TO SHIP** — 3 blockers verified closed via live probes on published wheels; interdiff confirms the rebase changed zero production code. |
| Full pytest suite | **13,489 passed, 0 failed** (10 skipped, 1 xfailed, 2 xpassed, 34 subtests); ran to completion serially (`-p no:xdist`, 9m31s). |
| Private/static gates | Threat scan CLEAN (score 0); ruff-forward / scope-undef clean; changed files compile. |
| My own source verification | Confirmed each of the 3 blocker fixes directly in the rebased worktree (below). |

### Findings — the three prior blockers, all reproduced CLOSED
- ✅ **[v0.18.0 static-skill capability gate] `api/profiles.py:80` + `api/streaming.py:7631`** — `_skill_modules_support_profile_home()` now requires both modules present with an unpatched `SKILLS_DIR`, a `_SKILLS_DIR_AT_IMPORT` baseline that still matches, AND a callable `_skills_dir()` that actually resolves the requested override. Fallback patching is skipped ONLY when the override installed AND both dynamic checks pass; otherwise the snapshot/patch/restore is retained. Both advisors reproduced: a named-profile turn on real v0.18.0 loads `alpha-only` (the original bounce returned `default-only`).
- ✅ **[checkpoint lock-ordering] `_save_streaming_checkpoint` `api/streaming.py:5401`** — now passes `scope_skill_modules=False`, so the checkpoint save never acquires the skill snapshot/patch lock (`Session.save()` doesn't touch skills). Lock order is uniformly patch-lock → `_ENV_LOCK`/`_agent_lock`, never reversed; a forced checkpoint during a static-fallback turn completed without hanging and still emitted `done` + `stream_end` (10/10 contention repeats).
- ✅ **[globally-patched capability false-positive] `api/profiles.py:109`** — the `current_skills_dir != import_skills_dir` check rejects globally-patched state as "dynamic". Overlapping alpha/beta turns starting from a globals-patched-to-alpha state no longer cross-leak (round-2 leak `alpha→beta-only` does not reproduce on either wheel).
- ✅ Restore-on-all-exits (success/exception/timeout) for home override, skill patch, and env is correct — `finally` under `_ENV_LOCK` with lock release in a nested `finally`.

### One finding — pre-existing master infra (does NOT block)
- ⚠️ Overlapping *dynamic-path* worker scopes can leave `os.environ['HERMES_HOME']` pointing at a named profile after both exit (interleaved save/restore of the shared env mirror). Both advisors reproduced it **byte-identically on pristine master** — it is not PR-introduced; the PR *eliminates* it for static agents (full-turn serialization) and routes dynamic-agent profile reads through the context-local override. **Follow-up (not this PR):** make the shared `HERMES_HOME` env-mirror save/restore generation-safe under overlap.

### Recommendation to the next agent
**Technically GREEN — all three maintainer blockers reproduced-closed against real wheels, full suite green, only a pre-existing-master concurrency artifact remains.** This is a profile-isolation / concurrency crown jewel on the streaming hot path, so it **parks for @nesquena's independent merge decision** (independent nesquena-account re-gate) rather than autonomous merge. Stage from `gate-b/5877-rebase` (patch-id-identical, current-master-based) with @starship-s attribution. File the env-mirror-overlap hardening as a follow-up.

---
_Gate-certifier layer (warm-up → **gate** → release). I do not merge/tag/deploy — that is the release agent's call. This cert is valid only while the head stays at sha:8838252bf0e8._

**Comment by nesquena-hermes** [comment]
## 🔍 Independent review requested — crown-jewel profile-isolation surface

@nesquena — this PR is **engineering-GREEN and converged**, but it lands on the concurrent-streaming / per-profile Hermes-home isolation crown jewel, so per project policy it needs your independent maintainer review (from the `nesquena` account, not the agent certifier) before it merges.

**State for your review:**
- **Exact head:** `8838252bf0e8` — MERGEABLE, CI 18/18 green.
- **Authoritative gate (certifier B, 2026-07-20):** Codex SAFE TO SHIP + Opus SAFE TO SHIP + full serial suite **13,489 passed / 0 failed** + threat scan CLEAN, on a patch-id-identical rebase (`f77a11d4`) onto current master.
- **All 3 prior maintainer blockers reproduced-closed against the real published v0.18.0/v0.18.1 skill wheels:**
  1. **v0.18.0 static-skill capability gate** — `_skill_modules_support_profile_home` (`profiles.py:80`) now requires unpatched `SKILLS_DIR == _SKILLS_DIR_AT_IMPORT` + a callable `_skills_dir()` that resolves the override, else retains snapshot/patch/restore (a v0.18.0 named turn loads alpha-only, not default-only).
  2. **Checkpoint lock-ordering** — `_save_streaming_checkpoint` (`streaming.py:5401`) passes `scope_skill_modules=False` so it never takes the skill lock; a forced checkpoint during a static-fallback turn completes + emits `done`/`stream_end` (10/10).
  3. **Globally-patched capability false-positive** — `profiles.py:109` rejects globally-patched state; overlapping alpha/beta no longer cross-leak.
- **One remaining finding (non-blocking):** the shared `HERMES_HOME` env-mirror save/restore can interleave under overlapping *dynamic-path* worker scopes — but this reproduces **byte-identically on pristine master** and is *improved* by this PR. Filed as a follow-up (make the env-mirror generation-safe), not a blocker for this PR.

**What I'd value your eyes on:** the per-turn profile-home scoping on the streaming hot path — specifically that the snapshot/patch/restore can't leak skills/config between two concurrent turns under different profiles, and that the checkpoint path's lock-avoidance is sound.

Once you sign off, I'll stage from the current-master-based patch-id-identical branch, preserve @starship-s attribution, run my own final live-head verification, and ship. Thanks @starship-s for the deep iteration here.

**Review by nesquena — APPROVED** [review]
## Review — end-to-end ✅ (independent maintainer review; clean approval, no fixes needed)

Independent review of the concurrent-profile Hermes-home isolation fix (@starship-s). This is the crown-jewel surface the agent certifier flagged for a `nesquena`-account review before merge — so I traced the isolation boundary, both closed blockers, and the teardown lifecycle myself against the real published agent.

### The bug

Concurrent chats can resolve different Hermes profiles in one WebUI process. The streaming path resolved the correct profile home, but **agent construction + later prompt/config/skill reads still depended on process-global `HERMES_HOME` and globally-patched skill-module paths after the env lock was released** — so a second profile's turn could overwrite those globals mid-first-turn and cross the profile boundary (one profile's `SOUL.md`/skills/config bleeding into another's chat, or a contaminated agent cached under a correct session identity).

### The fix — verified sound

**Context-local home, request-scoped for the whole turn.** `_run_agent_streaming` installs the agent's context-local override (`hermes_constants.set_hermes_home_override`, the same ContextVar `get_hermes_home()` consults first — I verified this primitive in #5577) *before* agent construction/MCP discovery/prompt load/tool exec, and resets it in the run's **terminal `finally`** (`streaming.py:10674`), with the ctx pre-initialized to a no-op triple (`(None, None, False)`) so a pre-install exception path resets safely. Because streaming turns run on their own worker threads and a ContextVar is per-thread, concurrent turns each hold their own override — genuine per-turn isolation, exception-safe. ✓

**The two maintainer blockers are closed in code, not just claimed:**
1. **Deadlock (checkpoint save):** `Session.save()`'s checkpoint path passes `scope_skill_modules=False` (`streaming.py:5232`), so a long-tool-turn checkpoint never takes the skill snapshot/patch lock → no `_agent_lock`→skill-lock ordering inversion. ✓
2. **Concurrency skill-leak (capability check):** `profiles.py:94-114` — a module is trusted as "dynamic" (global patching skipped) **only** if it still has `_SKILLS_DIR_AT_IMPORT` **and** `SKILLS_DIR == _SKILLS_DIR_AT_IMPORT`. A globally-patched module (the leak vector) fails this and is treated as static → forced into snapshot/patch/restore under the shared fallback lock. So a leaked-in profile path can never be trusted as the request's own. ✓ (Locked by `test_..._returns_false_when_already_globally_patched`.)

**Graceful degradation:** the override module is resolved lazily/optionally (`_resolve_streaming_hermes_home_override`); older agents without the context-local API fall through to the serialized snapshot/patch/restore fallback (held for the full turn), and the same isolation rule is applied to profile-scoped background workers. So the fix is safe on any agent version. ✓

**Scope discipline:** no SSE/transcript/replay/sidebar/API/UI change; env mirroring retained for legacy/subprocess readers. Confirmed the diff is streaming setup/teardown + profile-worker scope + tests only.

### Edge-case matrix

| Scenario | Behavior |
|---|---|
| Two concurrent turns, opposite profile homes | each resolves only its own home (context-local, per-thread) ✅ |
| Global `HERMES_HOME` clobbered mid-turn | reader still resolves the override profile ✅ (regression clobbers it deliberately) |
| Long-tool turn checkpoint while fallback lock held | completes — no skill-lock, no inversion ✅ |
| Skill module globally patched to another profile | capability check → static → fallback locking, not trusted ✅ |
| Older agent (no context-local API) | serialized snapshot/patch/restore fallback ✅ |
| Any streaming exit (error/cancel/normal) | override reset in terminal `finally` ✅ |

### Tests

- Focused profile/isolation suite (agent absent) — **186 passed, 2 skipped**; **with the real agent (PYTHONPATH=tarball) the isolation set is 19/19** (the concurrent-profile + real-`skills_list()` cases engage), incl. the 6 capability-check units (dynamic / globally-patched→false / static / raises / missing).
- Adjacent env-lock / profile-security — **12 passed**.
- Ruff forward-gate clean on changed lines. **Full suite: 13183 passed / 0 failed** in 317s (deselected the known darwin CRLF flake). CI **18/18 green**. Head = certified `8838252bf0e8`.

### Recommendation

✅ **Approved — independent maintainer review complete.** Parked at approval — ready for the release agent's merge/tag pipeline.

The isolation boundary is correct and exception-safe: the resolved profile home is bound through the agent's per-thread context-local override for the entire streaming turn (and detached worker scope) and reset in the terminal `finally`, so concurrent profiles can't cross even when process globals are clobbered mid-run. Both prior blockers are genuinely closed in the code I read — checkpoint saves skip the skill lock (no deadlock), and the capability check rejects globally-patched module state (no leak) — with a serialized fallback that keeps older agents safe. Converged through a two-round + rebase gate; my independent trace + full-suite/agent-present runs agree. Nice work, @starship-s. Ship.

**Comment by nesquena-hermes** [comment]
Both prior round-1/round-2 blockers from my earlier CHANGES_REQUESTED reviews are verified closed at the current head `8838252bf0e8` (checkpoint save skips the skill lock → no deadlock; capability check rejects globally-patched module state → no leak), and @nesquena has now completed an independent maintainer review from the human account with a clean end-to-end APPROVED (no fixes needed). Superseding my earlier requests — this is converged. Staging for release with @starship-s attribution; running a final full-suite verification on the current-master rebase (code patch-id identical to the approved head `2be850ee`) before merge.

**Comment by nesquena-hermes** [comment]
Shipped in **exp-v0.52.133** (via #6379). Concurrent-profile Hermes-home isolation is live — the resolved profile home is now bound through a per-thread context-local override for the whole streaming turn (+ detached worker scope) and reset in the terminal `finally`, so concurrent profiles can't cross even when process globals are clobbered mid-run. Independent maintainer review (@nesquena) APPROVED end-to-end; agent gate-pass + full suite clean at the rebased head (code patch-id identical to the approved head). Excellent, deep work @starship-s — thank you! 🎯
